@app/client/src/next.config.js

```diff
@@ -31,6 +38,12 @@ if (!ROOT_URL) {
       withCss,
       withLess
     )({
+      serverRuntimeConfig: {
+        BUCKET: BUCKET,
+        AWSACCESSKEYID: AWSACCESSKEYID,
+        AWSSECRETKEY: AWSSECRETKEY,
```

Comment:
This is extremely insecure: you are passing the AWS secret key to the browser! That means any user of your application can find your secret key, and make AWS API requests on your behalf! 😱 

A better, more secure way of doing this is to generate a signed upload URL on the server, and only pass that signed URL to the client. That way, the client has enough information to do what it needs to do (upload files to S3), but nothing more than that. Fortunately, there is a library called [`react-s3-uploader`](https://github.com/odysseyscience/react-s3-uploader) that is designed to work in exactly this way. I highly suggest using this library.

Reply:
Thanks for the code review! I've been programming solo so far, so its good to get feedback.  

This is the suggested way to pass server only [runtime configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) to Next.  

![image](https://user-images.githubusercontent.com/18180168/73594127-cb204f00-44d0-11ea-886b-3ec65e4a2d9c.png)  

I think..... that these are only available on the server side, since it's used only in a [next api route](https://nextjs.org/docs/api-routes/introduction).  

The intent is that this does exactly what you said, "...generate a signed upload URL on the server, and only pass that signed URL to the client".
I'll look into using the `react-s3-uploader` library.

Can you confirm if this is actually sending the serverRuntimeConfig: credentials to the client?? 
If so, I have some other things I need to change quickly 😳...
At least I'm using limited IAM roles.

Reply:
Oh dear, I'm sorry for the panic. After reading the Next.js docs more carefully, I think you're right: the credentials should stay on the server.

I saw that you put the credentials in the config, and then imported and used them in a file under the `/pages` directory -- and in my limited experience with Next.js, everything under `/pages` is run on both the server _and_ the client. But I didn't know that [`serverRuntimeConfig` is special](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) and [`/pages/api` is special](https://nextjs.org/docs/api-routes/introduction). (Seems weird to me that an API is a subset of a page...)

I tried to run the project locally, just to check for sure. Unfortunately, I ran into [a bug with graphql-codegen](https://github.com/dotansimha/graphql-code-generator/issues/3326) that prevented me from running it. However, you should be able to check yourself by [searching all of your client-side javascript in Chrome DevTools](https://stackoverflow.com/a/7970237/141395). If you search for your secret key (or even just for the string "AWSSECRETKEY", since you used that in the config), then you _should_ get no results.

Reply:
No result in the search when searching for the key or any of the other server side vars.
So, at least that part seems to be secure.
I still agree that the code quality could be better.